### PR TITLE
Hotfix rest Chinese encoding

### PR DESCRIPTION
**Pull request by MajaChen**
## What is the purpose of the change
When utilizing Dubbo REST with the @QueryParam annotation on parameters, issues with Chinese character decoding may arise. Refer to the [13349 issue](https://github.com/apache/dubbo/issues/13349#issuecomment-1879523852) for a comprehensive and detailed description of the problem, including its cause, solution, and an explanation of why it works.

## Brief changelog
RequestFacade.initParameters will decode the request uri with UTF-8 before parsing uri to parameters.

## Verifying this change
- I add more unit tests , then can pass, and I recommend you refer the [13349 issue](https://github.com/apache/dubbo/issues/13349#issuecomment-1879523852) 

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [x] Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [x] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [x] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [x] GitHub Actions works fine on your own branch.
- [x] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Comment by AlbumenJ** [comment]
@suncairong163 PTAL

**Comment by MajaChen** [comment]
I'm wondering if anybody can review my code

File: dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/request/RequestFacade.java [hunk 1]
```diff
@@ -56,6 +58,17 @@ protected void initHeaders() {}
     protected void initParameters() {
         String requestURI = getRequestURI();
 
+        String decodedRequestURI = requestURI;
+        try {
+            decodedRequestURI = URLDecoder.decode(decodedRequestURI, "UTF-8");
+        } catch (UnsupportedEncodingException e) {
+            // do nothing
+        }
```
Comment (CrazyHZM) [inline dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/request/RequestFacade.java]: This is hardcoded, we can set utf-8 by default, but it must come from `Accept-Charset` first.

**Review by CrazyHZM — CHANGES_REQUESTED** [review]

File: dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/request/RequestFacade.java (hunk 1 above)
Comment (MajaChen) [reply dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/request/RequestFacade.java]: ok, I'll make a adjustment

Comment (MajaChen) [reply dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/request/RequestFacade.java]: I've made an adjustment according to your opinon, please take a look again, take your time

**Comment by kimmking** [comment]
As json standard format, non-ansi chars are not allowed, so is nessary to do this?

**Comment by MajaChen** [comment]
Take an example of SpringMVC framework, Chinese character encoding problems are very common among Chinese developers, and it offers solution. I think dubbo should also offer such solution to avoid such problems, even though it's not common to pass Chinese character as parameter. But in my situation,I does need to do so.

**Comment by MajaChen** [comment]
And many users including me, are not familiar with  json standard format(sorry for that).So this bug can be a trap

**Comment by kimmking** [comment]
I know your case now.
BTW: It need/maybe process q param to parse Accept-Charset field like:
```
Accept-Charset: <charset>

// Multiple types, weighted with the [quality value](https://developer.mozilla.org/zh-CN/docs/Glossary/Quality_values) syntax:
Accept-Charset: utf-8, iso-8859-1;q=0.5
```

refer:
https://developer.mozilla.org/zh-CN/docs/Web/HTTP/Headers/Accept-Charset

**Comment by MajaChen** [comment]
I add a new method parseAcceptCharset, it can deal with multi values related with Accept-Charset.
It worked well as demonstrated by unit test.
I think it's ok to take the charset with the biggest weight alone.

I'm doubtful whether these code have pratical function in most case. As Accept-Charset is not recommended to use.
Most importantly, OkhttpClient don't take Accept-Charset into consideration when encoding url,see:
builder.url(requestTemplate.getURL());  (OKHttpRestClient.java line 55)

But I still think adding Accept-Charset make code more rigorous and compatible

**Comment by MajaChen** [comment]
PATL, thanks

File: dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/request/RequestFacade.java [hunk 2]
```diff
@@ -55,6 +58,24 @@ protected void initHeaders() {}
 
     protected void initParameters() {
         String requestURI = getRequestURI();
+        String decodedRequestURI = null;
+
+        try {
+            String enc = "UTF-8";
```
Comment (CrazyHZM) [inline dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/request/RequestFacade.java]: It should also be placed in `Constant`.

**Review by CrazyHZM — CHANGES_REQUESTED** [review]

File: dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/request/RequestFacade.java (hunk 2 above)
Comment (MajaChen) [reply dubbo-rpc/dubbo-rpc-rest/src/main/java/org/apache/dubbo/rpc/protocol/rest/request/RequestFacade.java]: fine

**Review by CrazyHZM — APPROVED** [review]

**Pull request merged**